File: sdk/python/studex/orchestrator.py
"""
Orchestrator API — Multi-agent coordination.

Pattern matches Google ADK's workflow orchestration:
    task = client.orchestrator.tasks.create("deploy the dashboard")
    print(task.plan)
    for a in task.assignments:
        print(f"  → {a.agent}: {a.task}")
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from studex._client import StudExClient
from studex._types import (
    OrchestratorStatus,
    TaskResult,
    TaskAssignment,
    BlackboardEntry,
)


class OrchestratorAPI:
    """
    Orchestrator API — Coordinate all 5 agents, route tasks, manage workflows.

    Agents can send messages to each other, share facts via the blackboard,
    and execute multi-step workflows through Hermes routing.
    """

    def __init__(self, client: StudExClient) -> None:
        self._client = client
        self.tasks = TaskAPI(client)
        self.blackboard = BlackboardAPI(client)
        self.workflows = WorkflowAPI(client)

    # ── Status ────────────────────────────────────────────────────────────

    def status(self) -> OrchestratorStatus:
        """Get orchestrator status snapshot."""
        data = self._client._get("/orchestrate")
        return OrchestratorStatus(
            agents=data.get("agents", 0),
            registered=data.get("registered", []),
            blackboard_keys=data.get("blackboardKeys", 0),
            messages_logged=data.get("messagesLogged", 0),
            active_workflows=data.get("activeWorkflows", 0),
            uptime=data.get("uptime", 0),
        )

    def agents(self) -> List[Dict[str, Any]]:
        """List all registered agents in the mesh."""
        return self._client._get("/orchestrate/agents")

    def message_log(self, count: int = 50) -> List[Dict[str, Any]]:
        """Get inter-agent message history."""
        data = self._client._get("/orchestrate/log", {"count": str(count)})
        return data.get("messages", [])

    # ── Agent-to-Agent (direct) ──────────────────────────────────────────

    def send(
        self,
        from_agent: str,
        to_agent: str,
        message: str,
    ) -> Dict[str, Any]:
        """Send a message from one agent to another."""
        return self._client._post("/orchestrate/send", {
            "from": from_agent,
            "to": to_agent,
            "message": message,
        })

    def broadcast(
        self,
        from_agent: str,
        message: str,
    ) -> Dict[str, Any]:
        """Broadcast a message to all agents."""
        return self._client._post("/orchestrate/broadcast", {
            "from": from_agent,
            "message": message,
        })

    def __repr__(self) -> str:
        return "<OrchestratorAPI>"


class TaskAPI:
    """Task routing via Hermes coordinator."""

    def __init__(self, client: StudExClient) -> None:
        self._client = client

    def create(self, task: str, from_agent: str = "user") -> TaskResult:
        """
        Route a multi-agent task through Hermes.

        Args:
            task: Task description (e.g. "deploy the dashboard")
            from_agent: Who is requesting (default: "user")

        Returns:
            TaskResult with plan, assignments, and results

        Example:
            task = client.orchestrator.tasks.create("deploy the dashboard")
            print(f"Plan: {task.plan}")
        """
        data = self._client._post("/orchestrate/task", {
            "from": from_agent,
            "task": task,
        })
        assignments = [
            TaskAssignment(
                agent=a.get("agent", ""),
                task=a.get("task", ""),
                response=a.get("response", ""),
                status=a.get("status", "completed"),
            )
            for a in data.get("assignments", [])
        ]
        return TaskResult(
            plan=data.get("plan", ""),
            assignments=assignments,
            results=data.get("results", {}),
        )

    def __repr__(self) -> str:
        return "<TaskAPI>"


class WorkflowAPI:
    """Multi-step workflow execution."""

    def __init__(self, client: StudExClient) -> None:
        self._client = client

    def create(
        self,
        name: str,
        steps: List[Dict[str, str]],
    ) -> Dict[str, Any]:
        """
        Execute a multi-step workflow across multiple agents.

        Args:
            name: Workflow name (e.g. "deploy-app")
            steps: List of {"agent": str, "task": str} dicts

        Example:
            client.orchestrator.workflows.create("deploy-app", [
                {"agent": "cursor", "task": "run tests"},
                {"agent": "openhuman", "task": "deploy to production"},
            ])
        """
        return self._client._post("/orchestrate/workflow", {
            "name": name,
            "steps": steps,
        })

    def __repr__(self) -> str:
        return "<WorkflowAPI>"


class BlackboardAPI:
    """Shared agent state (facts any agent can read/write)."""

    def __init__(self, client: StudExClient) -> None:
        self._client = client

    def get(self, key: str) -> Optional[BlackboardEntry]:
        """Get a fact from the blackboard."""
        try:
            data = self._client._get(f"/orchestrate/blackboard/{key}")
            if not data:
                return None
            return BlackboardEntry(
                key=key,
                value=data.get("value"),
                source=data.get("source", ""),
                updated_at=data.get("updatedAt", 0),
            )
        except Exception as e:
            if getattr(e, "status_code", None) == 404:
                return None
            raise

    def set(
        self,
        key: str,
        value: Any,
        source: str = "sdk",
    ) -> Dict[str, Any]:
        """Write a fact to the blackboard."""
        return self._client._post("/orchestrate/blackboard", {
            "key": key,
            "value": value,
            "source": source,
        })

    def delete(self, key: str) -> Dict[str, Any]:
        """Delete a fact from the blackboard."""
        return self._client._delete(f"/orchestrate/blackboard/{key}")

    def all(self) -> Dict[str, BlackboardEntry]:
        """Get all blackboard facts."""
        data = self._client._get("/orchestrate/blackboard")
        return {
            k: BlackboardEntry(
                key=k,
                value=v.get("value"),
                source=v.get("source", ""),
                updated_at=v.get("updatedAt", 0),
            )
            for k, v in data.items()
        }

    def __repr__(self) -> str:
        return "<BlackboardAPI>"

File: server/lib/orchestrator.js
/**
 * AgentOrchestrator — Inter-Agent Communication Hub
 *
 * Manages:
 *   - Agent registry (which agents are alive)
 *   - Shared blackboard (facts any agent can read/write)
 *   - Message routing between agents
 *   - Multi-agent workflow coordination
 *   - Broadcast to all agents
 *
 * Design: blackboard pattern + message bus.
 * Agents post facts, ask questions, and trigger workflows across each other.
 */
export default class AgentOrchestrator {
  constructor() {
    /** Map<agentName, BaseAgent> */
    this.agents = {};

    /** Shared blackboard — key-value store any agent can access */
    this.blackboard = new Map();

    /** Message log — history of inter-agent messages */
    this.messageLog = [];

    /** Workflow registry — multi-step coordinated tasks */
    this.workflows = new Map();

    this._startedAt = Date.now();
  }

  // ── Registry ──────────────────────────────────────────────────────────────

  /** Register an agent into the orchestration system. */
  register(agent) {
    // Prevent double-wrapping if register is called more than once
    if (this.agents[agent.name]) return;

    this.agents[agent.name] = agent;
    agent._orchestrator = this; // give agent a back-reference

    // Subscribe to agent events via a hook on the agent's chat method
    const originalChat = agent.chat.bind(agent);
    agent.chat = async (message) => {
      const response = await originalChat(message);
      // Log the interaction to the orchestrator
      this.messageLog.push({
        from: agent.name,
        to: '__self__',
        message,
        response,
        timestamp: new Date().toISOString(),
      });
      return response;
    };

    this.messageLog.push({
      from: 'system',
      to: agent.name,
      message: `Agent ${agent.name} registered`,
      response: 'acknowledged',
      timestamp: new Date().toISOString(),
    });
  }

  /** Get an agent by name. */
  get(name) {
    return this.agents[name] || null;
  }

  /** List all registered agents. */
  list() {
    return Object.values(this.agents).map(a => ({
      name: a.name,
      role: a.role,
      status: a.state.status,
      rvs: a.rvs,
    }));
  }

  // ── Blackboard (shared state) ─────────────────────────────────────────────

  /** Set a fact on the shared blackboard. */
  setFact(key, value, source) {
    this.blackboard.set(key, { value, source, updatedAt: Date.now() });
    this.messageLog.push({
      from: source || 'system',
      to: '__blackboard__',
      message: `set ${key} = ${typeof value === 'object' ? JSON.stringify(value) : value}`,
      response: 'stored',
      timestamp: new Date().toISOString(),
    });
  }

  /** Get a fact from the shared blackboard. */
  getFact(key) {
    return this.blackboard.get(key) || null;
  }

  /** Get all facts on the blackboard. */
  getAllFacts() {
    const facts = {};
    for (const [key, entry] of this.blackboard) {
      facts[key] = entry;
    }
    return facts;
  }

  /** Delete a fact. */
  deleteFact(key) {
    this.blackboard.delete(key);
  }

  // ── Inter-Agent Messaging ────────────────────────────────────────────────

  /**
   * Send a message from one agent to another.
   * Returns the receiving agent's response.
   */
  async send(from, to, message) {
    const target = this.agents[to];
    if (!target) {
      throw new Error(`Agent '${to}' not found in registry`);
    }

    const response = await target.chat(`[from ${from}] ${message}`);

    this.messageLog.push({
      from,
      to,
      message,
      response,
      timestamp: new Date().toISOString(),
    });

    return response;
  }

  /**
   * Route a task to the best agent(s) using Hermes as the router.
   */
  async routeTask(from, taskDescription) {
    const hermes = this.agents['hermes'];
    if (!hermes) {
      throw new Error('Hermes (router) not registered');
    }

    // Ask Hermes to determine which agents should handle this task
    const routing = await hermes.coordinate(taskDescription);

    const results = {};
    for (const assignment of routing.assignments) {
      const agent = this.agents[assignment.agent];
      if (agent) {
        const response = await agent.chat(`[task from ${from}] ${assignment.task}`);
        results[assignment.agent] = {
          task: assignment.task,
          response,
          status: 'completed',
        };
        this.messageLog.push({
          from: 'hermes',
          to: assignment.agent,
          message: assignment.task,
          response,
          timestamp: new Date().toISOString(),
        });
      }
    }

    return {
      plan: routing.explanation,
      assignments: routing.assignments,
      results,
    };
  }

  /**
   * Broadcast a message to all agents.
   */
  async broadcast(from, message) {
    const results = {};
    for (const [name, agent] of Object.entries(this.agents)) {
      if (name !== from) {
        try {
          const response = await agent.chat(`[broadcast from ${from}] ${message}`);
          results[name] = response;
          this.messageLog.push({
            from,
            to: name,
            message,
            response,
            timestamp: new Date().toISOString(),
          });
        } catch (err) {
          results[name] = `Error: ${err.message}`;
        }
      }
    }
    return results;
  }

  /**
   * Start a multi-step workflow.
   * Workflows are sequences of agent tasks executed in order.
   */
  async startWorkflow(name, steps) {
    const workflowId = `wf_${Date.now()}_${Math.random().toString(36).slice(2, 6)}`;
    const workflow = {
      id: workflowId,
      name,
      steps,
      currentStep: 0,
      startedAt: Date.now(),
      completed: false,
      results: [],
    };

    this.workflows.set(workflowId, workflow);

    // Execute steps sequentially
    for (let i = 0; i < steps.length; i++) {
      const step = steps[i];
      workflow.currentStep = i;
      const target = this.agents[step.agent];
      if (target) {
        const response = await target.chat(`[workflow: ${name}] ${step.task}`);
        workflow.results.push({
          step: i,
          agent: step.agent,
          task: step.task,
          response,
        });
        this.messageLog.push({
          from: 'workflow',
          to: step.agent,
          message: `[${name} step ${i + 1}/${steps.length}] ${step.task}`,
          response,
          timestamp: new Date().toISOString(),
        });
      }
    }

    workflow.completed = true;
    workflow.completedAt = Date.now();

    return {
      workflowId,
      name,
      steps: steps.length,
      results: workflow.results,
      completed: true,
    };
  }

  /** Get the message log (latest N entries). */
  getMessageLog(count = 50) {
    return this.messageLog.slice(-count);
  }

  /** Get uptime of the orchestrator. */
  get uptime() {
    return Math.floor((Date.now() - this._startedAt) / 1000);
  }

  /** Full orchestrator status snapshot. */
  get status() {
    return {
      agents: this.list().length,
      registered: Object.keys(this.agents),
      blackboardKeys: this.blackboard.size,
      messagesLogged: this.messageLog.length,
      activeWorkflows: [...this.workflows.values()].filter(w => !w.completed).length,
      uptime: this.uptime,
    };
  }
}
